sensor_board/common/pinav_parser.c

```diff
@@ -0,0 +1,266 @@
+			*i *= 10;
+			*i += *it - '0';
+		}
+		return SUCCESS;
+	}
+	return FAIL;
+}
+
+// Private helper function for parsing fix time from GGA sentence
+static generic_status_t parse_gga_fix_time(uint32_t * time, uint8_t * start) {
+	int32_t hours;
+	if (!stoi(&hours, start, 2)) {
+		return FAIL;
+	}
+	int32_t minutes;
+	if (!stoi(&minutes, start + 2, 2)) {
+		return FAIL;
+	}
+	int32_t seconds;
+	if (!stoi(&seconds, start + 4, 2)) {
+		return FAIL;
+	}
+	if (start[6] != '.') {
+		return FAIL;
+	}
+	int32_t millis;
+	if (!stoi(&millis, start + 7, 3)) {
+		return FAIL;
+	}
+
+	// Set out to the number of milliseconds since midnight UTC
+	*time = millis;
+	*time += (seconds * 1000);
+	*time += (minutes * 60000);
+	*time += (hours * 60 * 60000);
+	return SUCCESS;
+}
+
+// Private helper function for parsing out latitude from gga sentence
+// uint8_t * it should point to the start of the latitude field
+static generic_status_t parse_gga_lat(int32_t * lat, uint8_t * it) {
+	int32_t degrees;
+	if (!stoi((uint32_t *)&degrees, it, 2)) {
+		return FAIL;
+	}
+	int32_t minutes;
+	if (!stoi((uint32_t *)&minutes, it + 2, 2)) {
+		return FAIL;
+	}
+	if (it[4] != '.') {
+		return FAIL;
+	}
+	int32_t ten_thousandths_of_minute;
+	if (!stoi((uint32_t *)&ten_thousandths_of_minute, it + 5, 4)) {
+		return FAIL;
+	}
+
+	//calculate 600,000ths of a degree
+	int32_t six_100000s_of_degrees = degrees * 600000;
+	six_100000s_of_degrees += minutes * 10000;
+	six_100000s_of_degrees += ten_thousandths_of_minute;
+
+	// Scale to millionths of a degree
+	int32_t millionths_of_degrees = (six_100000s_of_degrees * 10) / 6;
+
+	// check N/S
+	if (it[9] != ',') {	// If this isn't safe, we would have failed by now
+		return FAIL;
+	}
+	if (it[10] == 'N') {
+		*lat = millionths_of_degrees;
+	}
+	else if (it[10] == 'S') {
+		*lat = 0 - millionths_of_degrees;
+	}
+	else {
+		return FAIL;
+	}
+
+	return SUCCESS;
+}
+
+// Private helper function for parsing out longitude from gga sentence
+// uint8_t * it should point to the start of the longitude field
+static generic_status_t parse_gga_long(int32_t * lng, uint8_t * it) {
+	int32_t degrees;
+	if (!stoi((uint32_t *)&degrees, it, 3)) {
+		return FAIL;
+	}
+	int32_t minutes;
+	if (!stoi((uint32_t *)&minutes, it + 3, 2)) {
+		return FAIL;
+	}
+	if (it[5] != '.') {
+		return FAIL;
+	}
+	int32_t ten_thousandths_of_minute;
+	if (!stoi((uint32_t *)&ten_thousandths_of_minute, it + 6, 4)) {
+		return FAIL;
+	}
+
+	//calculate 600,000ths of a degree
+	int32_t six_100000s_of_degrees = degrees * 600000;
+	six_100000s_of_degrees += minutes * 10000;
+	six_100000s_of_degrees += ten_thousandths_of_minute;
+
+	// Scale to millionths of a degree
+	int32_t millionths_of_degrees = (six_100000s_of_degrees * 10) / 6;
+
+	// check N/S
+	if (it[9] != ',') {	// If this isn't safe, we would have failed by now
+		return FAIL;
+	}
+	if (it[10] == 'E') {
+		*lng = millionths_of_degrees;
+	}
+	else if (it[10] == 'W') {
+		*lng = 0 - millionths_of_degrees;
```

Comment:
Why `0 - ` instead of unary `-`? This is actually a little less ambiguous but it's kinda odd =P.